src/python/pants/help/help_info_extracter.py

```diff
@@ -442,20 +492,31 @@ def get_rule_infos(
             for rule, providers in build_configuration.rule_to_providers.items()
             if isinstance(rule, TaskRule)
         ]
-        return {
-            rule_output_type: tuple(infos)
-            for rule_output_type, infos in groupby(
-                sorted(rule_infos, key=cls.rule_info_output_type), key=cls.rule_info_output_type
-            )
-        }
+
+        def _capture(t: Iterable[RuleInfo]) -> Callable[[], tuple[RuleInfo, ...]]:
+            data = tuple(t)
```

Comment:
I don't think I follow why this is outside of the inner function?

Reply:
It is because the iterable `infos` variable in the for loop that we get from `groupby` is only valid/usable until the next for loop iteration.

Reply:
Got it, that has been the source of a couple bugs for Pants. I think a comment would help.